pkg/apiserver/resthandler.go

```diff
@@ -488,8 +488,7 @@ func PatchResource(r rest.Patcher, scope RequestScope, typer runtime.ObjectTyper
 		gv := scope.Kind.GroupVersion()
 		codec := runtime.NewCodec(
 			scope.Serializer.EncoderForVersion(s, gv),
-			scope.Serializer.DecoderToVersion(s, unversioned.GroupVersion{Group: gv.Group, Version: runtime.APIVersionInternal}),
-		)
+			scope.Serializer.DecoderToVersion(s, gv))
```

Comment:
DecoderToVersion does conversions (if needed) to get from the kind/version represented in the JSON to the kind/version of the given type. you can prove this to yourself if you like by adding logging statements to something that only happens in conversion, and which deals with a field that only exists in the external struct (like copying a deprecated external field into an internal field):

https://gist.github.com/liggitt/2acaeaa8974b69db6e7a25dd6e1e668e

That example proves that the JSON is deserialized into the external struct (since the "serviceAccount" field doesn't exist internally), and that the conversion path is exercised to convert from the v1 struct to the internal struct.

With that logging in place, create a pod, then issue a PATCH API request:

```
curl -X PATCH http://localhost:8080/api/v1/namespaces/default/pods/mypod -H "Content-Type: application/strategic-merge-patch+json" -d '{"spec":{"serviceAccountName":"","serviceAccount":"default2"}}'
```

you'll see the conversion code getting called from the resthandler DecodeInto call:
```
k8s.io/kubernetes/pkg/api/v1.Convert_v1_PodSpec_To_api_PodSpec(0xc8213e8670, 0xc820cdb4f0, 0x7049ec0, 0xc821350dc0, 0x0, 0x0)
	/Users/jliggitt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/api/v1/conversion.go:473 +0x7da
k8s.io/kubernetes/pkg/api/v1.autoConvert_v1_Pod_To_api_Pod(0xc8213e8580, 0xc820cdb400, 0x7049ec0, 0xc821350dc0, 0x0, 0x0)
	/Users/jliggitt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/api/v1/conversion_generated.go:4388 +0x17d
k8s.io/kubernetes/pkg/api/v1.Convert_v1_Pod_To_api_Pod(0xc8213e8580, 0xc820cdb400, 0x7049ec0, 0xc821350dc0, 0x0, 0x0)
	/Users/jliggitt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/api/v1/conversion.go:580 +0x376
k8s.io/kubernetes/pkg/api/install.addVersionsToScheme.func1(0x33925e0, 0xc8213e8580, 0x3388de0, 0xc820cdb400, 0x7049ec0, 0xc821350dc0, 0x7000000, 0x0, 0x0)
	/Users/jliggitt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/api/install/install.go:164 +0x1038
k8s.io/kubernetes/pkg/conversion.(*Converter).Convert(0xc8200c0c80, 0x33925e0, 0xc8213e8580, 0x3388de0, 0xc820cdb400, 0x8, 0xc82023cf58, 0x0, 0x0)
	/Users/jliggitt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/conversion/converter.go:519 +0x19e
k8s.io/kubernetes/pkg/runtime.(*Scheme).Convert(0xc82006f2c0, 0x33925e0, 0xc8213e8580, 0x3388de0, 0xc820cdb400, 0x0, 0x0)
	/Users/jliggitt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/runtime/scheme.go:456 +0x71c
k8s.io/kubernetes/pkg/runtime.unsafeObjectConvertor.Convert(0xc82006f2c0, 0x33925e0, 0xc8213e8580, 0x3388de0, 0xc820cdb400, 0x0, 0x0)
	<autogenerated>:91 +0x6b
k8s.io/kubernetes/pkg/runtime/serializer/versioning.(*codec).Decode(0xc8212e90e0, 0xc82011d700, 0x55e, 0xa06, 0x0, 0x7093c68, 0xc820cdb400, 0x0, 0x0, 0xc8200cc800, ...)
	/Users/jliggitt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/runtime/serializer/versioning/versioning.go:163 +0x3c8
k8s.io/kubernetes/pkg/runtime.(*codec).Decode(0xc82191aca0, 0xc82011d700, 0x55e, 0xa06, 0x0, 0x7093c68, 0xc820cdb400, 0x0, 0x0, 0xc82051b440, ...)
	<autogenerated>:5 +0xca
k8s.io/kubernetes/pkg/runtime.DecodeInto(0x70554f0, 0xc82191aca0, 0xc82011d700, 0x55e, 0xa06, 0x7093c68, 0xc820cdb400, 0x0, 0x0)
	/Users/jliggitt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/runtime/codec.go:60 +0x98
k8s.io/kubernetes/pkg/apiserver.patchResource.func1(0x73213c0, 0xc8218b4990, 0x0, 0x0, 0x7093c68, 0xc820ef6000, 0x0, 0x0, 0x0, 0x0)
	/Users/jliggitt/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:577 +0x136a
```